pkg/lint/lintersdb/validator_test.go

```diff
@@ -0,0 +1,410 @@
+	},
+	{
+		desc: "disable-all and disable with enable",
+		cfg: &config.Linters{
+			Enable:     []string{"nolintlint"},
+			EnableAll:  false,
+			Disable:    []string{"dupl", "gofmt", "misspell"},
+			DisableAll: true,
+			Fast:       false,
+		},
+		expected: "can't combine options --disable-all and --disable dupl",
+	},
+	{
+		desc: "enable-all and enable",
+		cfg: &config.Linters{
+			Enable:     []string{"dupl", "gofmt", "misspell"},
+			EnableAll:  true,
+			Disable:    nil,
+			DisableAll: false,
+			Fast:       false,
+		},
+		expected: "can't combine options --enable-all and --enable dupl",
+	},
+}
+
+type validatorTestCase struct {
+	desc string
+	cfg  *config.Linters
+}
+
+var validateLintersNamesTestCases = []validatorTestCase{
+	{
+		desc: "no enable no disable",
+		cfg: &config.Linters{
+			Enable:  nil,
+			Disable: nil,
+		},
+	},
+	{
+		desc: "existing enabled linter",
+		cfg: &config.Linters{
+			Enable:  []string{"gofmt"},
+			Disable: nil,
+		},
+	},
+	{
+		desc: "existing disabled linter",
+		cfg: &config.Linters{
+			Enable:  nil,
+			Disable: []string{"gofmt"},
+		},
+	},
+}
+
+var validatePresetsTestCases = []validatorTestCase{
+	{
+		desc: "known preset",
+		cfg: &config.Linters{
+			EnableAll: false,
+			Presets:   []string{"bugs"},
+		},
+	},
+	{
+		desc: "enable-all and no presets",
+		cfg: &config.Linters{
+			EnableAll: true,
+			Presets:   nil,
+		},
+	},
+	{
+		desc: "no presets",
+		cfg: &config.Linters{
+			EnableAll: false,
+			Presets:   nil,
+		},
+	},
+}
+
+var validateDisabledAndEnabledAtOneMomentTestCases = []validatorTestCase{
+	{
+		desc: "2 different sets",
+		cfg: &config.Linters{
+			Enable:  []string{"dupl", "gofmt", "misspell"},
+			Disable: []string{"goimports", "gosec", "nolintlint"},
+		},
+	},
+	{
+		desc: "only enable",
+		cfg: &config.Linters{
+			Enable:  []string{"goimports", "gosec", "nolintlint"},
+			Disable: nil,
+		},
+	},
+	{
+		desc: "only disable",
+		cfg: &config.Linters{
+			Enable:  nil,
+			Disable: []string{"dupl", "gofmt", "misspell"},
+		},
+	},
+	{
+		desc: "no sets",
+		cfg: &config.Linters{
+			Enable:  nil,
+			Disable: nil,
+		},
+	},
+}
+
+var validateAllDisableEnableOptionsTestCases = []validatorTestCase{
+	{
+		desc: "nothing",
+		cfg:  &config.Linters{},
+	},
+	{
+		desc: "enable and disable",
+		cfg: &config.Linters{
+			Enable:     []string{"goimports", "gosec", "nolintlint"},
+			EnableAll:  false,
+			Disable:    []string{"dupl", "gofmt", "misspell"},
+			DisableAll: false,
+		},
+	},
+	{
+		desc: "disable-all and enable",
+		cfg: &config.Linters{
+			Enable:     []string{"goimports", "gosec", "nolintlint"},
+			EnableAll:  false,
+			Disable:    nil,
+			DisableAll: true,
+		},
+	},
+	{
+		desc: "enable-all and disable",
+		cfg: &config.Linters{
+			Enable:     nil,
+			EnableAll:  true,
+			Disable:    []string{"goimports", "gosec", "nolintlint"},
+			DisableAll: false,
+		},
+	},
+	{
+		desc: "enable-all and enable and fast",
```

Comment:
offtop: could you explain why `Fast` in this condition at all?

```
if cfg.EnableAll && len(cfg.Enable) != 0 && !cfg.Fast {
```

Reply:
Because of the way that fast is handled.
Cannot be changed for now: https://github.com/golangci/golangci-lint/issues/1909